script.js

```diff
@@ -0,0 +1,176 @@
+        return "'directories' property must be defined and be an object";
+    }
+
+    if (!packageJson.goBinary) {
+        return "'goBinary' property is required";
+    }
+
+    if (!packageJson.directories.bin) {
+        return "'directories.bin' property is required";
+    }
+}
+
+function parsePackageJson() {
+    if (!(process.arch in ARCH_MAPPING)) {
+        console.error("Installation is not supported for this architecture: " + process.arch);
+        return;
+    }
+
+    if (!(process.platform in PLATFORM_MAPPING)) {
+        console.error("Installation is not supported for this platform: " + process.platform);
+        return;
+    }
+
+    const packageJsonPath = path.join(".", "package.json");
+    if (!fs.existsSync(packageJsonPath)) {
+        console.error("Unable to find package.json. " + "Please run this script at root of the package you want to be installed");
+        return;
+    }
+
+    const packageJson = JSON.parse(fs.readFileSync(packageJsonPath));
+    const error = validateConfiguration(packageJson);
+    if (error && error.length > 0) {
+        console.error("Invalid package.json: " + error);
+        return;
+    }
+
+    // We have validated the config. It exists in all its glory
+    let binName = packageJson.goBinary;
+    let binNameUnused = binName + ".exe";
+    const binPath = packageJson.directories.bin;
+
+    // Binary name on Windows has .exe suffix
+    if (process.platform === "win32") {
+        binNameUnused = binName;
+        binName += ".exe";
+    }
+
+    return {
+        binName,
+        binNameUnused,
+        binPath
+    };
+}
+
+/**
+ * Reads the configuration from application's package.json,
+ * validates properties, copies the binary from the package and stores at
+ * ./bin in the package's root. NPM already has support to install binary files
+ * specific locations when invoked with "npm install -g"
+ *
+ *  See: https://docs.npmjs.com/files/package.json#bin
+ */
+
+function install(callback) {
+    const opts = parsePackageJson();
+    if (!opts) {
+        return callback("Invalid inputs");
+    }
+
+    const source = path.join('dist', `enonic_${PLATFORM_MAPPING[process.platform]}_${ARCH_MAPPING[process.arch]}`, opts.binName);
+
+    if (!fs.existsSync(source)) {
+        console.error('Downloaded binary does not contain the binary specified in configuration - ' + opts.binName);
+        return;
+    }
+
+    const targetPath = path.join(__dirname, opts.binPath);
+    const target = path.join(targetPath, opts.binName);
+
+    console.info(`Copying the relevant binary for your platform ${process.platform}`);
+    fs.copyFileSync(source, target, fs.constants.COPYFILE_FICLONE);
+}
+
+function preinstall(callback) {
+    const opts = parsePackageJson();
+    if (!opts) {
+        return callback("Invalid inputs");
+    }
+
+    const targetPath = path.join(__dirname, opts.binPath);
+    const unusedBinary = path.join(targetPath, opts.binNameUnused);
```

Comment:
`const unusedBinary = path.join(__dirname, opts.binPath, opts.binNameUnused);` ?

Reply:
Yeah, lines 102 and 103 can be optimized too.

Reply:
Done